API life cycle and deprecation policy in official documentation (#4529)
pytorch/executorch: docs/source/api-life-cycle.md

* _Deprecated_ APIs will be clearly marked (see the “How to Mark API State” section below).
* _Deprecated_ APIs will remain functional for at least the _deprecation period_ (see the “Deprecation Period” section below) to allow developers time to migrate to alternative APIs.

_Deleted_

* APIs whose removal are made permanent. Cleaned up from both code and documentation.

## Deprecation Policy

Follow these steps to deprecate and remove an API:

1. Discuss the change and collect initial feedback.
2. Clearly mark the API deprecated in code and documentation (See “How to Mark API State” below).
3. Listen to user feedback after the first release that deprecates the API. Users who weren't involved in the original discussion may have good arguments for not deprecating or removing the API.
4. Once the deprecation period has passed, the API may be removed (See “Deprecation Period” below). Be sure to also remove references from the documentation.

We also use deprecation as a way to make breaking changes to an existing interface: for example, if adding a non-optional parameter to a method. To do this without breaking existing users:
1. In a single commit:

>    - Create a new API that meets the new requirements.

Reviewer: Consistency: the rest of the doc uses `*` for bullets, but this uses `-`

   </td>
  </tr>
  <tr>
   <td>Python
   </td>
   <td>

<em>Deprecated</em>: use the <a href="https://typing-extensions.readthedocs.io/en/latest/#typing_extensions.deprecated">typing_extensions.deprecated</a> decorator

<em>Experimental</em>: use ExecuTorch's native experimental (TODO not yet implemented) decorator

   </td>
   <td>

Use <code>.. warning::</code> in the docstrings of deprecated and experimental APIs. See <a href="https://github.com/pytorch/pytorch/blob/cd8bbdc71a0258292381a7d54c8b353988d02ff4/torch/nn/utils/stateless.py#L170">example usage</a>.
</ul>
   </td>
  </tr>
  <tr>
   <td>C++
   </td>
   <td>

Use the <code>__ET_DEPRECATED</code> or (TODO not yet implemented) <code>__ET_EXPERIMENTAL</code> macros. See <a href="https://github.com/pytorch/executorch/blob/8e0f856ee269b319ac4195509cf31e3f548aa0e8/runtime/executor/program.h#L81">example usage</a>.
</ul>
   </td>
   <td>

Start Doxygen comments with "DEPRECATED:" or "EXPERIMENTAL:". See <a href="https://github.com/pytorch/executorch/blob/9d859653ae916d0a72f6b2b5c5925bed38832140/runtime/executor/program.h#L139">example usage</a>.

   </td>
  </tr>
  <tr>
   <td>Android
   </td>

>    <td>Deprecated: use <a href="https://docs.oracle.com/javase/9/docs/api/java/lang/Deprecated.html">java.lang.Deprecated</a>

Reviewer: Consistency: This uses non-italicized "Deprecated" and "Experimental", but the first row italicizes them. The C++ row doesn't even mention "Deprecated" or "Experimental" directly. And the Android/Apple rows only seem to talk about deprecated code, not experimental code.

Ideally, each row of this table would be consistent with all of the other rows.

<p>

> Experimental: use <a href="https://developer.android.com/reference/androidx/annotation/RequiresOptIn">androidx.annotation.RequiresOptIn</a>

Reviewer: Maybe point to https://cs.android.com/androidx/platform/frameworks/support/+/androidx-main:docs/api_guidelines/annotations.md as well (or instead), which give more guidance about using this annotation.

   </td>
   <td>/**
<p>
@deprecated As of release x.x, replaced by {@link #newMethod()}
<p>
*/
<p>
/**
<p>

> Warning: This API is experimental

Reviewer: Our style guide prefers full sentences, ending with periods.
```suggestion
Warning: This API is experimental.
```

<p>
*/
   </td>
  </tr>
  <tr>
   <td>Apple
   </td>
   <td>In Objective-C:
<p>
<code>__attribute__((deprecated("Use newMethod instead")));</code>
<p>
In Swift:
<p>
<code>@available(*, deprecated, message: "Use newMethod instead")</code>
   </td>
   <td>In Objective-C:
<p>

> /**
> <p>
> * @deprecated Use `newMethod` instead.
> <p>
> */

Reviewer: Would be nice to make these code blocks, though I don't know if that works in a table.
```suggestion
/**
<pre><code>
<p>
* @deprecated Use `newMethod` instead.
<p>
</code></pre>
*/
```

Or maybe triple-backquote will work here.

Same for the swift example

</li>
</ul>
   </td>
  </tr>
  <tr>
   <td><em>Deleted</em>
   </td>
   <td>
<ul>

<li>APIs whose removal are made permanent. Cleaned up from both code and documentation.
</li>
</ul>
   </td>
  </tr>
</table>

## Deprecation Policy

Follow these steps to deprecate an API:

1. Discuss the change and collect initial feedback.
2. Clearly mark the API deprecated in code and documentation (See “How to Mark?” below). The mark would trigger static and/or runtime warning that contains at least these information (or in documentation only, if a warning message is not feasible):
    1. Clearly point to the non-deprecated alternative to migrate to, or be clear if there is no alternative.
    2. Specify the earliest version in which the API may actually be removed (See “Deprecation Period” below)
3. Watch out for feedback. Users who weren’t involved in the original discussion can now comment. Maybe reconsider the deprecation.
4. The API removal may now be made permanent having reached the declared version. Make sure documentation is also cleaned up.

> In the case that a breaking change has to be made to an existing interface, for example, if a non-optional parameter has to be added to a method, then we treat this as a deprecation case. Steps to take:

Reviewer: This sentence is a little hard to follow. Suggested rewording.

```suggestion
We also use deprecation as a way to make breaking changes to an existing interface: for example, if adding a non-optional parameter to a method. To do this without breaking existing users:
```

> 1. Create a new API that meets the new requirements.
> 2. Migrate use cases from the old API to the new API.
> 3. Deprecate the old API.

Reviewer: Deprecation should happen at the same time as the introduction of the new API.

```suggestion
1. In a single commit:
  - Create a new API that meets the new requirements.
  - Deprecate the old API and recommend that users move to the new API.
2. Migrate use cases from the old API to the new API.
3. Delete the old API after the deprecation period.
```

## How to Mark?

> We want to make the best effort to mark in the most standard way possible, so that it’s easier for users to integrate with their dev tools, like Linters.

Reviewer: ```suggestion
When possible, the ExecuTorch code uses language-standard ways to annotate API lifecycle state in the code. This makes it easier for IDEs and other tools to communicate state to developers.
```

<table>
  <tr>
   <td><strong>API</strong>
   </td>
   <td><strong>Code</strong>
   </td>
   <td><strong>Documentation</strong>
   </td>
  </tr>
  <tr>
   <td>Python
   </td>
   <td>
<ul>

> <li><em>Deprecated</em>: use <a href="https://typing-extensions.readthedocs.io/en/latest/#typing_extensions.deprecated">typing_extensions.deprecated</a> decorator

Reviewer: ```suggestion
<li><em>Deprecated</em>: use the <a href="https://typing-extensions.readthedocs.io/en/latest/#typing_extensions.deprecated">typing_extensions.deprecated</a> decorator
```

<li><em>Experimental</em>: use ExecuTorch's native experimental (TODO not yet implemented) decorator
</li>
</ul>
   </td>
   <td>
<ul>

> <li>Use <code>.. warning::</code> in docstring of deprecated and experimental APIs, see example usage <a href="https://github.com/pytorch/pytorch/blob/cd8bbdc71a0258292381a7d54c8b353988d02ff4/torch/nn/utils/stateless.py#L170">here</a>.

Reviewer: ```suggestion
<li>Use <code>.. warning::</code> in the docstrings of deprecated and experimental APIs. See <a href="https://github.com/pytorch/pytorch/blob/cd8bbdc71a0258292381a7d54c8b353988d02ff4/torch/nn/utils/stateless.py#L170">example usage</a>.
```

</li>
</ul>
   </td>
  </tr>
  <tr>
   <td>C++
   </td>
   <td>
<ul>

> <li>Use the <code>__ET_DEPRECATED</code> and (TODO not yet implemented) <code>__ET_EXPERIMENTAL</code> macro, see example usage <a href="https://github.com/pytorch/executorch/blob/8e0f856ee269b319ac4195509cf31e3f548aa0e8/runtime/executor/program.h#L81">here</a>.

Reviewer: ```suggestion
<li>Use the <code>__ET_DEPRECATED</code> or (TODO not yet implemented) <code>__ET_EXPERIMENTAL</code> macros. See <a href="https://github.com/pytorch/executorch/blob/8e0f856ee269b319ac4195509cf31e3f548aa0e8/runtime/executor/program.h#L81">example usage</a>.
```

</li>
</ul>
   </td>
   <td>
<ul>

> <li>Start docstring with “DEPRECATED” and “EXPERIMENTAL”, see example usage <a href="https://github.com/pytorch/executorch/blob/9d859653ae916d0a72f6b2b5c5925bed38832140/runtime/executor/program.h#L139">here</a>.

Reviewer: ```suggestion
<li>Start Doxygen comments with "DEPRECATED:" or "EXPERIMENTAL:". See <a href="https://github.com/pytorch/executorch/blob/9d859653ae916d0a72f6b2b5c5925bed38832140/runtime/executor/program.h#L139">example usage</a>.
```

</li>
</ul>
   </td>
  </tr>
  <tr>
   <td>Android
   </td>
   <td>TODO
   </td>
   <td>TODO
   </td>
  </tr>
  <tr>
   <td>iOS
   </td>
   <td>TODO
   </td>
   <td>TODO
   </td>
  </tr>
</table>

## Deprecation Period

Here we recommend waiting for at least 2 minor releases before the removal. For example, if a function is marked as _deprecated_ in release 1.3.x, then it can be _deleted_ in 1.5.x or later.

> 
> 
> ## Future Enhancements
> 
> 
> ### Enforcement
> 
> We want to implement automated tests to monitor:
> * Were any BC-breaking changes made to stable APIs?
> * Did the removal of an deprecated API happen not sooner than the promised period?
> 
> 
> ### Auto migration
> 
> The current deprecation policy still requires users to take actions to migrate off from deprecated APIs. Ideally, we should automate this process. We want to leverage tools like [TorchFix](https://pypi.org/project/TorchFix/) to auto-migrate uses of deprecated APIs when possible.

Reviewer: Please remove this section. We do want to work on it, but it's not relevant to the policy itself.
```suggestion
```

<p>
<p>
Use <code>__ET_EXPERIMENTAL</code> macros (TODO not yet implemented)
</ul>
   </td>
   <td>

Start Doxygen comments with <code>DEPRECATED:</code> See
<a href="https://github.com/pytorch/executorch/blob/9d859653ae916d0a72f6b2b5c5925bed38832140/runtime/executor/program.h#L139">example
usage</a>

<p>
<p>
Start Doxygen comments with <code>EXPERIMENTAL:</code>
   </td>
  </tr>
  <tr>
   <td>Java
   </td>
   <td>

Use <a href="https://docs.oracle.com/javase/9/docs/api/java/lang/Deprecated.html">java.lang.Deprecated</a>

<p>
<p>

Use <a href="https://cs.android.com/androidx/platform/frameworks/support/+/androidx-main:docs/api_guidelines/annotations.md">androidx.annotation.RequiresOptIn</a>

   </td>
   <td>
<p>
   <code>/**
<p>
* @deprecated Use {@link #newMethod()} instead.
<p>
*/
<p>
</code>
<p>
<code>
/**
<p>
* Warning: This API is experimental.
<p>
*/</code>
   </td>
  </tr>
  <tr>
   <td>Objective-C
   </td>
   <td>
<p>
<code>__attribute__((deprecated("Use newMethod instead")));</code>
<p>
<p>

> <code>__attribute__((experimental("Use newMethod instead")));</code> (TODO not yet implemented)

Reviewer: I don't know if it's possible to implement an attribute like this; they're typically defined by the toolchain.

Author: I'll let @shoumikhin comment on this. I've chatted with him offline and IIUC, there's a mapping between the Objective-C annotations and the CPP macros. So if the CPP `__ET_DEPRECATED` and `__ET_EXPERIMENTAL` can be implemented, then we can implement these.

Reviewer 2: Oh I meant we can use the deprecated annotation for experental features too, since there's no dedicated experimental annotation available. Thus __ET_EXPERIMENTAL is gonna be defined the same as deprecated, but with a different error message.
